server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java

```diff
@@ -73,78 +103,172 @@ public String name() {
-                Collections.emptyList(),
-                Collections.emptyList()
-            );
-        }
-        List<HealthIndicatorImpact> impacts = Collections.singletonList(
-            new HealthIndicatorImpact(
-                NAME,
-                REPOSITORY_CORRUPTED_IMPACT_ID,
-                1,
-                String.format(
-                    Locale.ROOT,
-                    "Data in corrupted snapshot repositor%s %s may be lost and cannot be restored.",
-                    corrupted.size() > 1 ? "ies" : "y",
-                    limitSize(corrupted, 10)
-                ),
-                List.of(ImpactArea.BACKUP)
-            )
+        var repositoryHealthAnalyzer = new RepositoryHealthAnalyzer(
+            clusterState,
+            totalRepositories,
+            corrupted,
+            healthInfo.repositoriesInfoByNode()
         );
         return createIndicator(
-            RED,
-            createCorruptedRepositorySummary(corrupted),
-            verbose
-                ? new SimpleHealthIndicatorDetails(
-                    Map.of(
-                        "total_repositories",
-                        totalRepositories,
-                        "corrupted_repositories",
-                        corruptedRepositories,
-                        "corrupted",
-                        limitSize(corrupted, 10)
-                    )
-                )
-                : HealthIndicatorDetails.EMPTY,
-            impacts,
-            List.of(
-                new Diagnosis(
-                    CORRUPTED_REPOSITORY,
-                    List.of(new Diagnosis.Resource(Type.SNAPSHOT_REPOSITORY, limitSize(corrupted, maxAffectedResourcesCount)))
-                )
-            )
+            repositoryHealthAnalyzer.getHealthStatus(),
+            repositoryHealthAnalyzer.getSymptom(),
+            repositoryHealthAnalyzer.getDetails(verbose),
+            repositoryHealthAnalyzer.getImpacts(),
+            repositoryHealthAnalyzer.getDiagnoses(maxAffectedResourcesCount)
         );
     }
 
-    private static String createCorruptedRepositorySummary(List<String> corrupted) {
-        var message = new StringBuilder().append("Detected [").append(corrupted.size()).append("] corrupted snapshot repositories: ");
-        collectionToDelimitedStringWithLimit(corrupted, ",", "[", "].", 1024, message);
-        return message.toString();
+    /**
+     * Analyzer for the cluster's repositories health; aids in constructing a {@link HealthIndicatorResult}.
+     */
+    static class RepositoryHealthAnalyzer {
+        private final ClusterState clusterState;
+        private final int totalRepositories;
+        private final List<String> corruptedRepositories;
+        private final Map<String, List<String>> unknownRepositories;
+        private final Map<String, List<String>> invalidRepositories;
```

Comment:
Then I won't be able to list the correct nodes in the ` affected_resources` field for the different diagnoses (invalid vs. unknown). I initially made this a map because I was planning on showing exactly which repo was broken on which node (per broken type). So I think I'll refactor this to have
```java
Set<String> unknownRepositories;
Set<String> unknownNodes;
Set<String> invalidRepositories;
Set<String> invalidNodes;
```
That way, I can also remove the `...Count` fields as I can just retrieve the size of the sets.